File: LeetCode/53. Maximum Subarray.java
class Solution {
  public int maxSubArray(int[] nums) {
    int currMax = nums[0];
    int maxSoFar = nums[0];
    for (int i = 1; i <= nums.length - 1; i++) {
      currMax = Math.max(nums[i], currMax + nums[i]);
      maxSoFar = Math.max(maxSoFar, currMax);
    }
    return maxSoFar;
  }
}

/**
TODO: Follow up: If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
**/

File: LeetCode/53. Maximum Subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let maxSoFar = -Infinity;
	let currMax = 0;

	for(let num of nums) {
		currMax = currMax + num;
		if(currMax < num) {
			currMax = num;
		}
		maxSoFar = Math.max(maxSoFar, currMax);
	}
	return maxSoFar;
};

/* Solution similar to Java submission */

var maxSubArray = function(nums) {
    let maxSoFar = -Infinity;
	let currMax = -Infinity;

	for(let num of nums) {
		currMax = Math.max(num, currMax + num);
		maxSoFar = Math.max(maxSoFar, currMax);
	}
	return maxSoFar;
};
